### PR TITLE
CI: Mark s3 tests parallel safe

### LINKED ISSUE #35856: CI: specified bucket does not exist in TestParquetPyArrow.test_s3_roundtrip_explicit_fs
might be related to #35655

```
================================== FAILURES ===================================
______________ TestParquetPyArrow.test_s3_roundtrip_explicit_fs _______________
[gw2] win32 -- Python 3.8.5 C:\Miniconda\envs\pandas-dev\python.exe

self = <pandas.tests.io.test_parquet.TestParquetPyArrow object at 0x000002B24268BF10>
df_compat =    A    B
0  1  foo
1  2  foo
2  3  foo
s3_resource = s3.ServiceResource(), pa = 'pyarrow'
s3so = {'client_kwargs': {'endpoint_url': 'http://127.0.0.1:5555/'}}

    def test_s3_roundtrip_explicit_fs(self, df_compat, s3_resource, pa, s3so):
        s3fs = pytest.importorskip("s3fs")
        if LooseVersion(pyarrow.__version__) <= LooseVersion("0.17.0"):
            pytest.skip()
        s3 = s3fs.S3FileSystem(**s3so)
        kw = dict(filesystem=s3)
>       check_round_trip(
            df_compat,
            pa,
            path="pandas-test/pyarrow.parquet",
            read_kwargs=kw,
            write_kwargs=kw,
        )

        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id,
                operation_name=operation_name),
            model=operation_model, params=request_dict,
            request_signer=self._request_signer, context=request_context)
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context)
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id,
                operation_name=operation_name),
            http_response=http, parsed=parsed_response,
            model=operation_model, context=request_context
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.NoSuchBucket: An error occurred (NoSuchBucket) when calling the PutObject operation: The specified bucket does not exist
```

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/35856

I think we need to update the pytest pattern though, so this should
fail.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
> so this should fail.

Does this mean that this PR should not be merged despite being green?

**Comment by TomAugspurger** [comment]
I'm going to rerun it a few times. I suspect we'll get a failure.

**Comment by TomAugspurger** [comment]
Updated to just make these fixtures parallel safe. At the start of each session, each worker starts up a moto server on `localhost:555{worker_id}`. Then we just need to plumb that through everywhere.

I think Martin mentioned possible memory issues on windows workers with this approach, but that may have been starting the moto server at the test level. Hopefully doing things at the session level avoids that.

**Comment by TomAugspurger** [comment]
Rerunning azure at https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40904&view=results. But we should be good.

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/pull/35655 wasn't backported, so this shouldn't need a backport either.

**Comment by TomAugspurger** [comment]
I've seen a couple failures on re-runs, but nothing to do with s3 / moto. So this should be good to go.

File: pandas/tests/io/conftest.py [hunk 1]
```diff
@@ -34,12 +34,13 @@ def feather_file(datapath):
 
 
 @pytest.fixture
-def s3so():
-    return dict(client_kwargs={"endpoint_url": "http://127.0.0.1:5555/"})
+def s3so(worker_id):
```
Comment (jbrockmendel) [inline pandas/tests/io/conftest.py:37]: @TomAugspurger where is this defined?  I'm getting test failures locally, guessing i need to update some dependency

Comment (TomAugspurger) [reply pandas/tests/io/conftest.py:37]: `worker_id` is from pytest-xdist.

**Comment by martindurant** [comment]
> I think Martin mentioned possible memory issues on windows workers

Yes, this was for the case of restarting the process for every test (i.e., standard pytest scope) - I believe moto may be less than optimal at cleaning up on exit, at least on Windows.

File: pandas/tests/io/conftest.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/io/conftest.py:37]: test_s3_roundtrip_for_dir is failing in the CI and im trying to reproduce locally but it is getting pytest.skipped and it isn't clear why.  could worker_id be involved?

**Pull request merged**